Support building cross arch via qemu

smoser opened the pull request:
The qemu runner "just works" for cross arch, you just have to supply it a target-arch kernel.  So the changes here really only amount to creating kernel/<ARCH>/vmlinuz rather than kernel/vmlinuz.

Now, tested inside a 'make dev-container' environment and also on an x86_64 Ubuntu 24.04 host, I can:

    make MELANGE_RUNNER=qemu ARCH=aarch64 debug/dash

And also:

    make MELANGE_RUNNER=qemu ARCH=aarch64 test-debug/dash

smoser committed:
Support building cross arch via qemu

The qemu runner "just works" for cross arch, you just have
to supply it a target-arch kernel.  So the changes here really
only amount to creating kernel/<ARCH>/vmlinuz rather than
kernel/vmlinuz.

Now, tested inside a 'make dev-container' environment and also
on an x86_64 Ubuntu 24.04 host, I can:

    make MELANGE_RUNNER=qemu ARCH=aarch64 debug/dash

And then also:

    make MELANGE_RUNNER=qemu ARCH=aarch64 test/dash

smoser commented:
Related https://github.com/chainguard-dev/melange/pull/1970

dannf commented on Makefile:
```diff
@@ -103,31 +104,32 @@ lib-token: ${CACHEDIR}/.libraries_token.txt
-	tar -x -C kernel -f $< boot/ 2> /dev/null
+kernel/%/vmlinuz: kernel/%/linux.apk
+	tmpd=kernel/.$$$$ && mkdir -p $$tmpd $(dir $@) && tar -x -C $$tmpd -f $< boot/ 2> /dev/null && \
+		mv $$tmpd/boot/* $(dir $@) && rc=$$?; rm -Rf $$tmpd; exit $$rc
```
maybe we could avoid the tmpd with `tar --strip-components=1` ?

```
dann-frazier@dann-frazier-laptop:~/git/os$ find kernel/x86_64/
kernel/x86_64/
kernel/x86_64/linux.apk
dann-frazier@dann-frazier-laptop:~/git/os$ tar -x -C kernel/x86_64/ -f kernel/x86_64/linux.apk boot/vmlinuz --strip-components=1
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
tar: Ignoring unknown extended header keyword 'APK-TOOLS.checksum.SHA1'
dann-frazier@dann-frazier-laptop:~/git/os$ find kernel/x86_64/
kernel/x86_64/
kernel/x86_64/vmlinuz
kernel/x86_64/linux.apk
```

smoser replied:
i dont have strong feelings either way. its a tmpdir and it gets cleaned up.

smoser replied:
I just added another change which i think actually justifies the tmpdir. Now we verify that we actually downloaded a vmlinuz rather than just touch-updated a potentially old one.

dustinkirkland commented:
This is sweet!  Thanks for banging this out so quickly!

dustinkirkland commented:
Works for me here on a Linux machine.  Would be great if someone could try this from a Mac and see if it can cross-compile x86 as well?

justinvreeland commented:
Tested on my Mac seems to be working fine

smoser committed:
Add QEMU_KERNEL_DEP to test deps, fix kernel target.

QEMU_KERNEL_DEP should be added to the test deps so you can
just 'make MELANGE_RUNNER=qemu test/dash' without having
to first 'make fetch-kernel'.

Second thing here is a fix to verify that a 'vmlinuz' is
created by the tar extraction.  Without that change
if the apk did not contain boot/vmlinuz, we'd created it with
'touch'
